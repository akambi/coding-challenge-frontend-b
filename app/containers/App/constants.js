export const QUEBEC = 'quebec';
export const MONTREAL = 'montreal';
export const API_KEY = 'PARTNER_BaASYYHxTxuOINEOMWq5GA';

export const CityToGeoHashMapping = {
  quebec: 'f2m673',
  montreal: 'f25dvk',
};

export const CHANGE_SEARCH_PARAMS = 'busbud/App/CHANGE_SEARCH_PARAMS';

export const LOAD_DEPARTURES = 'busbud/App/LOAD_DEPARTURES';
export const LOAD_DEPARTURES_SUCCESS = 'busbud/App/LOAD_DEPARTURES_SUCCESS';
export const LOAD_DEPARTURES_ERROR = 'busbud/App/LOAD_DEPARTURES_ERROR';
