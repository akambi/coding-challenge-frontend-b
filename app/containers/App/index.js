import React from 'react';
import { Helmet } from 'react-helmet';
import styled from 'styled-components';
import { Switch, Route } from 'react-router-dom';

import HomePage from 'containers/HomePage/Loadable';
import NotFoundPage from 'containers/NotFoundPage/Loadable';
import Header from 'components/Header';
import Footer from 'components/Footer';

import GlobalStyle from '../../global-styles';

const AppWrapper = styled.div`
  max-width: calc(768px + 16px * 2);
  margin: 0 auto;
  display: flex;
  min-height: 100%;
  padding: 0 16px;
  flex-direction: column;
`;

export default function App() {
  return (
    <AppWrapper>
      <Helmet
        titleTemplate="%s - Busbud's bus departures"
        defaultTitle="Busbud's bus departures"
      >
        <meta
          name="description"
          content="A simple application that lists all departures for a given origin city and a given destination city for a given day"
        />
      </Helmet>
      <Header />
      <Switch>
        <Route exact path="/" component={HomePage} />
        <Route path="" component={NotFoundPage} />
      </Switch>
      <Footer />
      <GlobalStyle />
    </AppWrapper>
  );
}
