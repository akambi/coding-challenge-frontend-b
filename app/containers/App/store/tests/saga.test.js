/**
 * Tests for App sagas
 */

import { delay, put, takeLatest } from 'redux-saga/effects';

import { LOAD_DEPARTURES, MONTREAL, QUEBEC } from 'containers/App/constants';
import {
  loadDepartures,
  departuresLoaded,
  departureLoadingError,
} from 'containers/App/store/actions';

import busbudData, { getDepartures } from '../saga';

const searchParams = {
  origin: QUEBEC,
  destination: MONTREAL,
  outboundDate: '2020-12-14T14:21:30.490Z',
  adult: 1,
};

/* eslint-disable redux-saga/yield-effects */
describe('getDepartures Saga', () => {
  let getDeparturesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    window.Headers = jest.fn();

    getDeparturesGenerator = getDepartures(loadDepartures());

    const selectDescriptor = getDeparturesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getDeparturesGenerator.next(searchParams).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the departuresLoaded action if it requests the data successfully and complete is true', () => {
    const response = {
      complete: true,
      locations: [
        {
          id: 1,
          name: QUEBEC,
        },
        {
          id: 2,
          name: MONTREAL,
        },
      ],
      departures: [
        {
          departure_time: '2020-12-13 09:10:00',
          arrival_time: '2020-12-13 14:00:00',
          origin_location_id: 1,
          prices: {
            total: 5520,
          },
        },
        {
          departure_time: '2020-12-13 09:10:00',
          arrival_time: '2020-12-13 14:00:00',
          origin_location_id: 2,
          prices: {
            total: 4920,
          },
        },
      ],
    };

    const departures = response.departures.map(departure => ({
      departure_time: departure.departure_time,
      arrival_time: departure.arrival_time,
      price: departure.prices.total,
      location: response.locations.find(
        location => location.id === departure.origin_location_id,
      ),
    }));

    const putDescriptor = getDeparturesGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(departuresLoaded(departures)));
  });

  it('should dispatch the departuresLoaded action if it requests the data successfully and completed is false', () => {
    const response = {
      complete: false,
      locations: [],
      departures: [],
      prices: {},
    };

    const delayDescriptor = getDeparturesGenerator.next(response).value;
    expect(delayDescriptor).toEqual(delay(2 * 1000));

    const putDescriptor = getDeparturesGenerator.next().value;
    expect(putDescriptor).toEqual(put(loadDepartures(true)));

    const returnDescriptor = getDeparturesGenerator.next().value;
    expect(returnDescriptor).toEqual(undefined);
  });

  it('should call the departureLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getDeparturesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(departureLoadingError(response)));
  });
});

describe('busbudDataSaga Saga', () => {
  const busbudDataSaga = busbudData();

  it('should start task to watch for LOAD_DEPARTURES action', () => {
    const takeLatestDescriptor = busbudDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(LOAD_DEPARTURES, getDepartures),
    );
  });
});
