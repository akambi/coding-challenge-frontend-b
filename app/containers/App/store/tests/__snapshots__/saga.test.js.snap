// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getDepartures Saga should call the departureLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getDepartures Saga should call the departureLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://napi.busbud.com/x-departures/f2m673/f25dvk/2020-12-14?adult=1",
      Object {
        "cache": "default",
        "headers": mockConstructor {},
        "method": "GET",
        "mode": "cors",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getDepartures Saga should dispatch the departuresLoaded action if it requests the data successfully and complete is true 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getDepartures Saga should dispatch the departuresLoaded action if it requests the data successfully and complete is true 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://napi.busbud.com/x-departures/f2m673/f25dvk/2020-12-14?adult=1",
      Object {
        "cache": "default",
        "headers": mockConstructor {},
        "method": "GET",
        "mode": "cors",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`getDepartures Saga should dispatch the departuresLoaded action if it requests the data successfully and completed is false 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`getDepartures Saga should dispatch the departuresLoaded action if it requests the data successfully and completed is false 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "https://napi.busbud.com/x-departures/f2m673/f25dvk/2020-12-14?adult=1",
      Object {
        "cache": "default",
        "headers": mockConstructor {},
        "method": "GET",
        "mode": "cors",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
